README.md
# RedisCache

A [Guava LoadingCache](https://google.github.io/guava/releases/22.0/api/docs/com/google/common/cache/LoadingCache.html)
backed by Redis.

This project is written in [Kotlin](https://kotlinlang.org/) and the examples in this file also use
Kotlin syntax.  While different from Java, Kotlin's syntax doesn't stray too far so it should be
intelligible to those who only know Java.  Similarly, Kotlin's Java interop is great and I have
tried to write code that can be [easily used in Java](https://android.jlelse.eu/writing-java-friendly-kotlin-code-c408b24fb4e).
If you find shortcomings in that respect, let me know!

Additionally, I encourage anyone who uses Java to give Kotlin a try—it's better in every way.

# Dependency

Find the latest version from the [repository page on Bintray](https://bintray.com/dinomite/maven/net.dinomite%3Aredis-cache).

Use that version number to add the dependency to your `build.gradle`:

    compile 'net.dinomite:redis-cache:[latest_version]'

Or `pom.xml`, if you use Maven:

    <dependency>
      <groupId>net.dinomite</groupId>
      <artifactId>redis-cache</artifactId>
      <version>[latest_version]</version>
    </dependency>

# Usage

The only thing RedisCache needs is a [JedisPool](https://github.com/xetorthio/jedis/wiki/Getting-started#using-jedis-in-a-multithreaded-environment)
for interacting with Redis.  Thereafter, it acts just like an Guava LoadingCache, except values are
stored in Redis.[1]

    // Pretend that this is expensive
    fun generateValue(key: String): String { return "value-for-$key" }
    
    val jedisPool = JedisPool("localhost", 6379)
    val redisCache = RedisCache<String, String>(jedisPool)
    redisCache.put("foo", { generateValue(String) })

You can optionally pass a `String` prefix that will be prepended to all keys.  RedisCache does not
add any separator between the prefix and individual keys; provide your own in the prefix string if
desired.

    val redisCache = RedisCache<String, String>(jedisPool, prefix = "page-cache:")

By default RedisCache uses the prefix 'redis-cache:'.

RedisCache sets TTL to 1 hour by default—no automatic eviction is performed so **you probably want a
different value**.  Pass the `expiration` parameter to set the TTL for all RedisCache-inserted
objects.

## Custom serialization

By default, RedisCache serializes both values & keys with the JDK's [ObjectOutputStream](https://docs.oracle.com/javase/8/docs/api/java/io/ObjectOutputStream.html)
written as a byte array. This works for anything you give it, but the keys produced are inscrutible
(well, for strings it's teh object type preceding the string value):

    127.0.0.1:6379> keys *
    1) "\xac\xed\x00\x05t\x00\x06foobar"

You can provide RedisCache with custom serializers for keys & values as constructor arguments.  If
you want your keys to be liegibile in Redis, you can use ToStringSerializer, which uses the provided
key object's `.toString()` method for serialization.  You probably only want this if you are using
strings, or other objects with unique & representative `.toString()` values.

    val redisCache = RedisCache(jedisPool = jedisPool, keySerializer = ToStringSerializer())
    ...
    127.0.0.1:6379> keys *
    1) "foobar"

RedisCache sends all keys & values to Redis as byte arrays.  You can implement the `Serializer`
interface and pass your implementation for key and/or value serialization.

# Development

## Tests

Tests are performed against [embedded-redis](https://github.com/kstyrc/embedded-redis).
`RedisCacheTest` tests all methods for a single-argument (i.e. default serializers, no prefix
or expiration, no loader) constructed instance.  The other test classes do what they say on the
tin, testing custom serializers & a default loader.  I split those out because they require
constructing a different RedisCache instance, and only need to test a subset of methods.

# TODO

- Smarter GetAll
- Refresh TTL on access
- Transactions for multi-host support?
- Implement LRU eviction

# See also

RedisCache started as a port from [this Java implementation](https://github.com/levyfan/guava-cache-redis)

[1]: The intent is to allow the cached values to be shareable amongst multiple machines, but
I have yet to actually try that.  Likely some transactions are needed in the RedisCache code.

CHANGELOG.md
# 0.2.0 - 2017-08-09
- Added expireAfterRead setting to allow updating of TTLs on every read of a
key


src/test/kotlin/net/dinomite/cache/RedisCacheTest.kt
package net.dinomite.cache

import com.google.common.cache.CacheLoader
import com.google.common.primitives.Bytes
import net.dinomite.cache.serializers.ObjectStreamSerializer
import org.junit.After
import org.junit.Assert.*
import org.junit.Before
import org.junit.Test
import redis.clients.jedis.JedisPool
import redis.embedded.RedisServer
import java.time.Duration
import java.util.*

class RedisCacheTest {
    val port = 16379
    val redisServer: RedisServer = RedisServer(port)
    val jedisPool = JedisPool("localhost", port)

    lateinit var redisCache: RedisCache<String, String>

    val prefix = "redis-cache:".toByteArray()
    val serializer = ObjectStreamSerializer()

    @Before
    fun setup() {
        redisServer.start()

        redisCache = RedisCache(jedisPool)
    }

    @After
    fun cleanup() {
        jedisPool.resource.use { it.flushDB() }
    }

    @Test
    fun testGetIfPresent_NullForNonExistentKey() {
        val key = "foobar"
        assertNull(redisCache.getIfPresent(key))
    }

    @Test
    fun testGetIfPresent_ReturnsValueThatExists() {
        val key = "foobar"
        val value = "foovalue"
        redisCache.put(key, value)
        assertEquals(value, redisCache.getIfPresent(key))
    }

    @Test
    fun testGet_WithLoader_LoadsValue() {
        val key = "foobar"
        val value = "foovalue"
        assertEquals(value, redisCache.get(key, { value }))
    }

    @Test
    fun testGet_WithLoader_ReturnsExistingValue() {
        val key = "foobar"
        val value = "foovalue"
        redisCache.put(key, value)
        assertEquals(value, redisCache.get(key, { value + "foobar" }))
    }

    @Test
    fun testGet_WithLoader_ExceptionForNullFromValueLoader() {
        val key = "foobar"
        try {
            redisCache.get(key, { null })
            fail("Expected CacheLoader.InvalidCacheLoadException")
        } catch (e: CacheLoader.InvalidCacheLoadException) {
            assertEquals("Loader must not return null, key=foobar", e.message)
        }
    }

    @Test
    fun testGetAllPresent_GetsAll() {
        val firstKey = "foo"
        val firstValue = "bar"
        val secondKey = "baz"
        val secondValue = "qux"
        val expected = mapOf(firstKey to firstValue, secondKey to secondValue)
        redisCache.putAll(expected)

        assertEquals(expected, redisCache.getAllPresent(listOf(firstKey, secondKey)))
    }

    @Test
    fun testGetAllPresent_SkipsMissing() {
        val firstKey = "foo"
        val firstValue = "bar"
        val secondKey = "baz"
        val secondValue = "qux"
        val expected = mapOf(firstKey to firstValue, secondKey to secondValue)
        redisCache.putAll(expected)

        assertEquals(expected, redisCache.getAllPresent(listOf(firstKey, secondKey, "quxx")))
    }

    @Test
    fun testPut() {
        val key = "foobar"
        val value = "foovalue"
        redisCache.put(key, value)
        assertEquals(value, redisCache.getIfPresent(key))
    }

    @Test
    fun testPut_UsesPrefix() {
        val key = "foobar"
        val value = "foovalue"
        redisCache.put(key, value)

        jedisPool.resource.use { jedis ->
            val serializedKey = Bytes.concat(prefix, serializer.serialize(key))
            val serializedValue = serializer.serialize(value)
            assertTrue(Arrays.equals(serializedValue, jedis.get(serializedKey)))
        }
    }

    @Test
    fun testPut_SetsExpiration() {
        val key = "foobar"
        val value = "foovalue"
        redisCache.put(key, value)

        jedisPool.resource.use { jedis ->
            assertEquals(Duration.ofHours(1).seconds, jedis.ttl(redisCache.buildKey(key)))
        }
    }

    @Test
    fun testPutAll() {
        val firstKey = "foo"
        val firstValue = "bar"
        val secondKey = "baz"
        val secondValue = "qux"
        val expected = mapOf(firstKey to firstValue, secondKey to secondValue)

        redisCache.putAll(expected)

        assertEquals(expected, redisCache.getAllPresent(listOf(firstKey, secondKey)))
    }

    @Test
    fun testInvalidate() {
        val key = "foobar"
        val value = "foovalue"
        redisCache.put(key, value)

        redisCache.invalidate(key)

        assertNull(redisCache.getIfPresent(key))
    }

    @Test
    fun testInvalidateAll() {
        val firstKey = "foo"
        val firstValue = "bar"
        val secondKey = "baz"
        val secondValue = "qux"
        val expected = mapOf(firstKey to firstValue, secondKey to secondValue)
        redisCache.putAll(expected)
        val keys = listOf(firstKey, secondKey)

        redisCache.invalidateAll(keys)

        assertTrue(redisCache.getAllPresent(keys).isEmpty())
    }

    @Test
    fun testGet_ThrowsExceptionForMissingLoader() {
        try {
            redisCache.get("foobar")
            fail("Expected IllegalStateException")
        } catch (e: IllegalStateException) {
            // Expected
        }
    }

    @Test
    fun testRefresh_ThrowsExceptionForMissingLoader() {
        try {
            redisCache.refresh("foobar")
            fail("Expected IllegalStateException")
        } catch (e: IllegalStateException) {
            // Expected
        }
    }

    @Test
    fun testBuildKey_IncludesPrefix() {
        val key = "foobar"
        val expected = Bytes.concat(prefix, serializer.serialize(key))
        val actual = redisCache.buildKey(key)
        assertTrue(Arrays.equals(expected, actual))
    }
}

src/test/kotlin/net/dinomite/cache/RedisCacheTest_WithLoader.kt
package net.dinomite.cache

import com.google.common.cache.CacheLoader
import org.junit.After
import org.junit.Assert.*
import org.junit.Before
import org.junit.Test
import redis.clients.jedis.JedisPool
import redis.embedded.RedisServer

class RedisCacheTest_WithLoader {
    val port = 16379
    val redisServer: RedisServer = RedisServer(port)
    val jedisPool = JedisPool("localhost", port)

    lateinit var redisCache: RedisCache<String, String>

    @Before
    fun setup() {
        redisServer.start()

        redisCache = RedisCache(jedisPool, loader = object: CacheLoader<String, String>() {
            override fun load(key: String): String {
                return "value-for-$key"
            }
        })
    }

    @After
    fun cleanup() {
        jedisPool.resource.use { it.flushDB() }
    }

    @Test
    fun testGetIfPresent_NullForNonexistantKey() {
        val key = "foobar"
        assertNull(redisCache.getIfPresent(key))
    }

    @Test
    fun testGetIfPresent_ReturnsValueThatExists() {
        val key = "foobar"
        val value = "foovalue"
        redisCache.put(key, value)
        assertEquals(value, redisCache.getIfPresent(key))
    }

    @Test
    fun testGet_WithLoader_LoadsValue() {
        val key = "foobar"
        val value = "foovalue"
        assertEquals(value, redisCache.get(key, { value }))
    }

    @Test
    fun testGet_WithLoader_ReturnsExistingValue() {
        val key = "foobar"
        val value = "foovalue"
        redisCache.put(key, value)
        assertEquals(value, redisCache.get(key, { value + "foobar" }))
    }

    @Test
    fun testGet_WithLoader_InsertsCustomValue() {
        val key = "foobar"
        try {
            redisCache.get(key, { null })
            fail("Expected CacheLoader.InvalidCacheLoadException")
        } catch (e: CacheLoader.InvalidCacheLoadException) {
            // Expected
        }
    }

    @Test
    fun testGetAllPresent_GetsAll() {
        val firstKey = "foo"
        val firstValue = "bar"
        val secondKey = "baz"
        val secondValue = "qux"
        val expected = mapOf(firstKey to firstValue, secondKey to secondValue)
        redisCache.putAll(expected)

        assertEquals(expected, redisCache.getAllPresent(listOf(firstKey, secondKey)))
    }

    @Test
    fun testGetAllPresent_SkipsMissing() {
        val firstKey = "foo"
        val firstValue = "bar"
        val secondKey = "baz"
        val secondValue = "qux"
        val expected = mapOf(firstKey to firstValue, secondKey to secondValue)
        redisCache.putAll(expected)

        assertEquals(expected, redisCache.getAllPresent(listOf(firstKey, secondKey, "quxx")))
    }

    @Test
    fun testPut() {
        val key = "foobar"
        val value = "foovalue"
        redisCache.put(key, value)
        assertEquals(value, redisCache.getIfPresent(key))
    }

    @Test
    fun testPutAll() {
        val firstKey = "foo"
        val firstValue = "bar"
        val secondKey = "baz"
        val secondValue = "qux"
        val expected = mapOf(firstKey to firstValue, secondKey to secondValue)

        redisCache.putAll(expected)

        assertEquals(expected, redisCache.getAllPresent(listOf(firstKey, secondKey)))
    }

    @Test
    fun testInvalidate() {
        val key = "foobar"
        val value = "foovalue"
        redisCache.put(key, value)

        redisCache.invalidate(key)

        assertNull(redisCache.getIfPresent(key))
    }

    @Test
    fun testInvalidateAll() {
        val firstKey = "foo"
        val firstValue = "bar"
        val secondKey = "baz"
        val secondValue = "qux"
        val expected = mapOf(firstKey to firstValue, secondKey to secondValue)
        redisCache.putAll(expected)
        val keys = listOf(firstKey, secondKey)

        redisCache.invalidateAll(keys)

        assertTrue(redisCache.getAllPresent(keys).isEmpty())
    }

    @Test
    fun testGet_LoadsValueWithLoader() {
        assertEquals("value-for-foo", redisCache.get("foo"))
    }

    @Test
    fun testRefresh_LoadsValueWithLoader() {
        redisCache.put("foo", "custom value")
        redisCache.refresh("foo")
        assertEquals("value-for-foo", redisCache.get("foo"))
    }
}

src/test/kotlin/net/dinomite/cache/RedisCacheTest_WithPrefix.kt
package net.dinomite.cache

import com.google.common.primitives.Bytes
import net.dinomite.cache.serializers.ObjectStreamSerializer
import org.junit.After
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test
import redis.clients.jedis.JedisPool
import redis.embedded.RedisServer
import java.util.*

class RedisCacheTest_WithPrefix {
    val port = 16379
    val redisServer: RedisServer = RedisServer(port)
    val jedisPool = JedisPool("localhost", port)

    lateinit var redisCache: RedisCache<String, String>

    val prefix = "prefix-"
    val serializer = ObjectStreamSerializer()

    @Before
    fun setup() {
        redisServer.start()

        redisCache = RedisCache(jedisPool, keyPrefix = prefix.toByteArray())
    }

    @After
    fun cleanup() {
        jedisPool.resource.use { it.flushDB() }
    }

    @Test
    fun testPut_UsesPrefix() {
        val key = "foobar"
        val value = "foovalue"
        redisCache.put(key, value)

        jedisPool.resource.use { jedis ->
            val serializedKey = Bytes.concat(prefix.toByteArray(), serializer.serialize(key))
            val serializedValue = serializer.serialize(value)
            assertTrue(Arrays.equals(serializedValue, jedis.get(serializedKey)))
        }
    }

    @Test
    fun testBuildKey_IncludesPrefix() {
        val key = "foobar"
        val expected = Bytes.concat(prefix.toByteArray(), serializer.serialize(key))
        val actual = redisCache.buildKey(key)
        assertTrue(Arrays.equals(expected, actual))
    }
}

src/main/kotlin/net/dinomite/cache/serializers/ObjectStreamSerializer.kt
package net.dinomite.cache.serializers

import java.io.*

/**
 * Serialize & deserialize objects to byte streams using the JDK's
 * ObjectOutputStream and ObjectInputStream.
 */
open class ObjectStreamSerializer : Serializer {
    override fun serialize(obj: Any?): ByteArray {
        val baos = ByteArrayOutputStream(512)
        try {
            ObjectOutputStream(baos).use { out -> out.writeObject(obj as Serializable) }
        } catch (ex: IOException) {
            throw SerializationException(ex)
        }
        return baos.toByteArray()
    }

    @Suppress("UNCHECKED_CAST", "Gotta happen ¯\\_(ツ)_/¯ ")
    override fun <T> deserialize(objectData: ByteArray): T {
        try {
            ObjectInputStream(ByteArrayInputStream(objectData)).use {
                return it.readObject() as T
            }
        } catch (ex: ClassNotFoundException) {
            throw SerializationException(ex)
        } catch (ex: IOException) {
            throw SerializationException(ex)
        }
    }
}


src/main/kotlin/net/dinomite/cache/serializers/SerializationException.kt
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.dinomite.cache.serializers

/**
 *
 * Exception thrown when the Serialization process fails.

 *
 * The original error is wrapped within this one.

 *
 * #NotThreadSafe# because Throwable is not thread-safe
 * @since 1.0
 */
class SerializationException : RuntimeException {

    /**
     *
     * Constructs a new `SerializationException` without specified
     * detail message.
     */
    constructor() : super() {}

    /**
     *
     * Constructs a new `SerializationException` with specified
     * detail message.

     * @param msg  The error message.
     */
    constructor(msg: String) : super(msg) {}

    /**
     *
     * Constructs a new `SerializationException` with specified
     * nested `Throwable`.

     * @param cause  The `Exception` or `Error`
     * *  that caused this exception to be thrown.
     */
    constructor(cause: Throwable) : super(cause) {}

    /**
     *
     * Constructs a new `SerializationException` with specified
     * detail message and nested `Throwable`.

     * @param msg    The error message.
     * *
     * @param cause  The `Exception` or `Error`
     * *  that caused this exception to be thrown.
     */
    constructor(msg: String, cause: Throwable) : super(msg, cause) {}

    companion object {

        /**
         * Required for serialization support.

         * @see java.io.Serializable
         */
        private val serialVersionUID = 4029025366392702726L
    }

}


src/main/kotlin/net/dinomite/cache/RedisCache.kt
package net.dinomite.cache

import com.google.common.annotations.VisibleForTesting
import com.google.common.cache.AbstractLoadingCache
import com.google.common.cache.CacheLoader
import com.google.common.cache.LoadingCache
import com.google.common.collect.ImmutableMap
import com.google.common.collect.Iterables
import com.google.common.primitives.Bytes
import net.dinomite.cache.serializers.ObjectStreamSerializer
import net.dinomite.cache.serializers.Serializer
import redis.clients.jedis.Jedis
import redis.clients.jedis.JedisPool
import redis.clients.jedis.Pipeline
import java.time.Duration
import java.util.*
import java.util.concurrent.Callable

/**
 * A Redis-backed [LoadingCache](https://google.github.io/guava/releases/22.0/api/docs/com/google/common/cache/LoadingCache.html)
 *
 *     val jedisPool = JedisPool("localhost", 6379)
 *     val redisCache = RedisCache<String, String>(jedisPool)
 *     redisCache.put("foo", { generateValue(String) })
 *
 * All keys are inserted with a expiration (1 hour unless specified) and prefix (the byte array representation of
 * "redis-cache:").  Setting `expireAfterRead` will refresh each key's expiration upon read.
 */
class RedisCache<K, V>
@JvmOverloads constructor(private val jedisPool: JedisPool,
                          private val keySerializer: Serializer = ObjectStreamSerializer(),
                          private val valueSerializer: Serializer = ObjectStreamSerializer(),
                          private val keyPrefix: ByteArray = "redis-cache:".toByteArray(),
                          expiration: Duration = Duration.ofHours(1),
                          private val loader: CacheLoader<K, V>? = null,
                          private val expireAfterRead: Boolean = false)
    : AbstractLoadingCache<K, V>(), LoadingCache<K, V> {

    private val expiration: Int = expiration.seconds.toInt()

    override fun getIfPresent(key: Any): V? {
        var ret: V? = null

        pipeline { p ->
            val locator = buildKey(key)
            val reply = p.get(locator)
            if (expireAfterRead) p.expire(locator, expiration)
            p.sync()

            val result = reply.get()
            if (result != null) {
                ret = valueSerializer.deserialize(result)
            }
        }

        return ret
    }

    @Throws(CacheLoader.InvalidCacheLoadException::class)
    override fun get(key: K, valueLoader: Callable<out V>): V {
        var value = this.getIfPresent(key as Any)
        if (value == null) {
            value = valueLoader.call()
            if (value == null) {
                throw CacheLoader.InvalidCacheLoadException("Loader must not return null, key=" + key)
            } else {
                this.put(key, value)
            }
        }

        return value
    }

    @Suppress("UNCHECKED_CAST", "Handles generic objects")
    override fun getAllPresent(keys: Iterable<Any?>): ImmutableMap<K, V> {
        var ret: Map<K, V> = mapOf()

        pipeline { p ->
            val locators = keys.map { buildKey(it) }
            val valueBytes = p.mget(*Iterables.toArray(locators, ByteArray::class.java))
            p.sync()

            val result = LinkedHashMap<K, V>()
            keys.map { it as K }.forEachIndexed { i, castKey ->
                val value = valueBytes.get()[i]
                if (value != null) {
                    if (expireAfterRead) p.expire(locators[i], expiration)
                    result.put(castKey, valueSerializer.deserialize(value))
                }
            }

            ret = result
        }

        return ImmutableMap.copyOf(ret)
    }

    override fun getAll(keys: MutableIterable<K>?): ImmutableMap<K, V> {
        // TODO Do getAllPresent(), then use pipeline to load remaining values
        return super.getAll(keys)
    }

    override fun put(key: K?, value: V?) {
        val keyBytes = buildKey(key)
        val valueBytes = valueSerializer.serialize(value)

        jedis { jedis ->
            jedis.setex(keyBytes, expiration, valueBytes)
        }
    }

    @Suppress("IMPLICIT_CAST_TO_ANY", "Handles generic objects")
    override fun putAll(items: Map<out K, V>) {
        val keysAndValues = ArrayList<ByteArray>()
        items.forEach { key, value ->
            keysAndValues.add(buildKey(key))
            keysAndValues.add(valueSerializer.serialize(value))
        }

        pipeline { p ->
            p.mset(*Iterables.toArray(keysAndValues, ByteArray::class.java))

            var i = 0
            while (i < keysAndValues.size) {
                p.expire(keysAndValues[i], expiration)
                i += 2
            }
            p.sync()
        }
    }

    override fun invalidate(key: Any?) {
        jedis { jedis ->
            jedis.del(buildKey(key))
        }
    }

    override fun invalidateAll(keys: Iterable<*>) {
        val keyBytes = keys.map { buildKey(it) }

        jedis { jedis ->
            jedis.del(*Iterables.toArray(keyBytes, ByteArray::class.java))
        }
    }

    override fun get(key: K): V {
        if (loader == null) {
            throw IllegalStateException("Cannot use single-argument get with null loader (provide one Cache construction)")
        }

        return this.get(key, { loader.load(key) })
    }

    override fun refresh(key: K) {
        if (loader == null) {
            throw IllegalStateException("Cannot refresh wih null loader (provide one Cache construction)")
        }

        this.put(key, loader.load(key))
    }

    /**
     * Serialize the given key & add the prefix
     */
    @VisibleForTesting
    internal fun buildKey(key: Any?): ByteArray {
        checkNotNull(key) { "Key cannot be null" }
        return Bytes.concat(keyPrefix, keySerializer.serialize(key))
    }

    private fun <T> jedis(body: (jedis: Jedis) -> T): T = jedisPool.resource.use { body(it) }

    private fun pipeline(body: (pipeline: Pipeline) -> Unit) {
        jedisPool.resource.use {
            it.pipelined().use {
                body(it)
                it.sync()
            }
        }
    }
}